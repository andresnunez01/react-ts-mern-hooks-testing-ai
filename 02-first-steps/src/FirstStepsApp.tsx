import { ItemCounter } from './shopping-cart/ItemCounter';

interface ItemInCart {
    productName: string;
    quantity: number;
}

const itemsInCart: ItemInCart[] = [
    { productName: 'Xbox Series X', quantity: 1},
    { productName: 'Control Xbox Series', quantity: 2},
    { productName: 'Hollow Knight: Silksong', quantity: 5},
]
export function FirstStepsApp() {
    return (
        <div data-testid="FSA-div">
            <h1>Carrito de compras</h1>

            {
                itemsInCart.map( ({productName, quantity}) => (
                    <ItemCounter key={productName} name={productName} quantity={quantity}/>
                ))
            }
            {/* <ItemCounter name="PlayStation 5" quantity={2}/>
            <ItemCounter name="Xbox Series X" quantity={3}/>
            <ItemCounter name="Hollow Knight: Silksong" quantity={8}/> */}

        </div>
    )
};