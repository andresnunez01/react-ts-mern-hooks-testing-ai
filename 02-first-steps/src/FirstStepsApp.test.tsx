import { render, screen } from "@testing-library/react";
import { beforeEach, describe, expect, test, vi } from "vitest";
import { FirstStepsApp } from "./FirstStepsApp";

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const mockItemCounter = vi.fn((_props: unknown) => {
    return (<div data-testid="itemCounter"/>)
})
    
vi.mock('./shopping-cart/ItemCounter', () => ({
    ItemCounter: (props: unknown) => mockItemCounter(props),
}))

// vi.mock('./shopping-cart/ItemCounter', ()=>({
//     ItemCounter: (props: unknown) => <div data-testid="itemCounter" name={props.name} quantity={props.quantity}/>
// }));

describe('FirstStepsApp', () => {
    beforeEach(()=>{
        vi.clearAllMocks();
    })
    test('should match snapshot', () => {

        render(<FirstStepsApp/>)

        expect(screen.getByTestId('FSA-div')).toMatchSnapshot();

        //screen.debug();
    });

    test('should render the correct number of itemCounter components', () => {
        render(<FirstStepsApp/>);

        //screen.debug();

        const itemCounters = screen.getAllByTestId('itemCounter');

        expect(itemCounters.length).toBe(3);

    })

    test('should render itemCounter with correct props', () => {
        render(<FirstStepsApp/>)

        expect(mockItemCounter).toHaveBeenCalledTimes(3);
        expect(mockItemCounter).toHaveBeenCalledWith({ name: 'Xbox Series X', quantity: 1});
        expect(mockItemCounter).toHaveBeenCalledWith({ name: 'Control Xbox Series', quantity: 2});
        expect(mockItemCounter).toHaveBeenCalledWith({ name: 'Hollow Knight: Silksong', quantity: 5});

    })
})