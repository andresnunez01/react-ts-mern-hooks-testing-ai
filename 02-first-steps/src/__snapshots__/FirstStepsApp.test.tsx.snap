// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FirstStepsApp > should match snapshot 1`] = `
<div
  data-testid="FSA-div"
>
  <h1>
    Carrito de compras
  </h1>
  <div
    data-testid="itemCounter"
  />
  <div
    data-testid="itemCounter"
  />
  <div
    data-testid="itemCounter"
  />
</div>
`;
